File: sdks/dotnet/sulfone-helium/Domain/Service/StatelessInquirer.cs
using System.Globalization;
using sulfone_helium.Domain.Core;
using sulfone_helium.Domain.Core.Questions;

namespace sulfone_helium.Domain.Service;

public class StatelessInquirer(Dictionary<string, IAnswer> answers) : IInquirer
{
    private readonly Dictionary<string, IAnswer> _answers = answers;

    private IAnswer GetAnswer(IQuestion q)
    {
        if (!_answers.ContainsKey(q.Id))
        {
            throw new OutOfAnswerException("", q);
        }

        return _answers[q.Id];
    }

    public Task<string[]> Checkbox(CheckboxQ q)
    {
        var answer = this.GetAnswer(q);
        var a = answer switch
        {
            StringArrayAnswer stringArrayAnswer => stringArrayAnswer.Answer,
            _ => throw new ArgumentOutOfRangeException(
                "Incorrect answer type. Expected: StringArrayAnswer. Got: " + answer.GetType()
            ),
        };
        return Task.FromResult(a);
    }

    public Task<string[]> Checkbox(string q, string[] options, string id, string? help)
    {
        return this.Checkbox(
            new CheckboxQ()
            {
                Message = q,
                Options = options,
                Desc = help,
                Id = id,
            }
        );
    }

    public Task<bool> Confirm(ConfirmQ q)
    {
        var answer = this.GetAnswer(q);
        var a = answer switch
        {
            BoolAnswer ans => ans.Answer,
            _ => throw new ArgumentOutOfRangeException(
                "Incorrect answer type. Expected: BoolAnswer. Got: " + answer.GetType()
            ),
        };
        return Task.FromResult(a);
    }

    public Task<bool> Confirm(string q, string id, string? help)
    {
        return this.Confirm(
            new ConfirmQ()
            {
                Message = q,
                Desc = help,
                Id = id,
            }
        );
    }

    public Task<string> Password(PasswordQ q)
    {
        var answer = this.GetAnswer(q);
        var a = answer switch
        {
            StringAnswer ans => ans.Answer,
            _ => throw new ArgumentOutOfRangeException(
                "Incorrect answer type. Expected: StringAnswer. Got: " + answer.GetType()
            ),
        };
        return Task.FromResult(a);
    }

    public Task<string> Password(string q, string id, string? help)
    {
        return this.Password(
            new PasswordQ()
            {
                Message = q,
                Desc = help,
                Id = id,
            }
        );
    }

    public Task<string> Select(SelectQ q)
    {
        var answer = this.GetAnswer(q);
        var a = answer switch
        {
            StringAnswer ans => ans.Answer,
            _ => throw new ArgumentOutOfRangeException(
                "Incorrect answer type. Expected: StringAnswer. Got: " + answer.GetType()
            ),
        };
        return Task.FromResult(a);
    }

    public Task<string> Select(string q, string[] options, string id, string? help)
    {
        return this.Select(
            new SelectQ()
            {
                Message = q,
                Desc = help,
                Options = options,
                Id = id,
            }
        );
    }

    public Task<string> Text(TextQ q)
    {
        var answer = this.GetAnswer(q);
        var a = answer switch
        {
            StringAnswer ans => ans.Answer,
            _ => throw new ArgumentOutOfRangeException(
                "Incorrect answer type. Expected: StringAnswer. Got: " + answer.GetType()
            ),
        };
        return Task.FromResult(a);
    }

    public Task<string> Text(string q, string id, string? help)
    {
        return this.Text(
            new TextQ()
            {
                Message = q,
                Desc = help,
                Id = id,
            }
        );
    }

    public Task<DateOnly> DateSelect(DateQ q)
    {
        var answer = this.GetAnswer(q);
        var a = answer switch
        {
            StringAnswer ans => ans.Answer,
            _ => throw new ArgumentOutOfRangeException(
                "Incorrect answer type. Expected: StringAnswer. Got: " + answer.GetType()
            ),
        };
        var d = DateOnly.ParseExact(a, "yyyy-MM-dd", CultureInfo.InvariantCulture);
        return Task.FromResult(d);
    }

    public Task<DateOnly> DateSelect(string question, string id, string? help)
    {
        return this.DateSelect(
            new DateQ()
            {
                Message = question,
                Desc = help,
                Id = id,
            }
        );
    }
}

File: sdks/node/src/domain/service/stateless_inquirer.ts
import type { IInquirer } from '../core/inquirer.js';
import type { Answer } from '../core/answer.js';
import { isBoolAnswer, isStringAnswer, isStringArrayAnswer } from '../core/answer.js';
import { OutOfAnswerException } from './out_of_answer_error.js';
import type { CheckboxQ, ConfirmQ, DateQ, PasswordQ, Question, SelectQ, TextQ } from '../core/question.js';
import { QuestionType } from '../core/question.js';

class StatelessInquirer implements IInquirer {
  readonly #answers: Record<string, Answer>;

  constructor(answers: Record<string, Answer>) {
    this.#answers = answers;
  }

  private getAnswer(q: Question): Answer {
    if (!this.#answers[q.id]) {
      throw new OutOfAnswerException('', q);
    }

    return this.#answers[q.id];
  }

  checkbox(q: string, options: string[], id: string, help?: string | null): Promise<string[]>;
  checkbox(q: CheckboxQ): Promise<string[]>;
  async checkbox(q: CheckboxQ | string, options?: string[], id?: string, help?: string | null): Promise<string[]> {
    if (typeof q === 'string') {
      if (options == null) {
        throw new Error('options cannot be null');
      }
      if (id == null) {
        throw new Error('id cannot be null');
      }
      return this.checkbox({
        message: q,
        options,
        desc: help,
        type: QuestionType.Checkbox,
        id,
      });
    }
    const answer = this.getAnswer(q);
    if (isStringArrayAnswer(answer)) return Promise.resolve(answer.answer);

    throw new Error('Incorrect answer type. Expected: StringArrayAnswer. Got: ' + typeof answer);
  }

  confirm(q: ConfirmQ): Promise<boolean>;
  confirm(q: string, id: string, help?: string | null): Promise<boolean>;

  async confirm(q: ConfirmQ | string, id?: string, help?: string | null): Promise<boolean> {
    if (typeof q === 'string') {
      if (id == null) {
        throw new Error('id cannot be null');
      }
      return this.confirm({
        message: q,
        desc: help,
        type: QuestionType.Confirm,
        id,
      });
    }
    const answer = this.getAnswer(q);

    if (isBoolAnswer(answer)) return Promise.resolve(answer.answer);
    throw new Error('Incorrect answer type. Expected: BoolAnswer. Got: ' + typeof answer);
  }

  password(q: string, id: string, help?: string | null): Promise<string>;
  password(q: PasswordQ): Promise<string>;
  async password(q: PasswordQ | string, id?: string, help?: string | null): Promise<string> {
    if (typeof q === 'string') {
      if (id == null) {
        throw new Error('id cannot be null');
      }
      return this.password({
        type: QuestionType.Password,
        message: q,
        desc: help,
        id,
      });
    }
    const answer = this.getAnswer(q);

    if (isStringAnswer(answer)) return Promise.resolve(answer.answer);
    throw new Error('Incorrect answer type. Expected: StringAnswer. Got: ' + typeof answer);
  }

  select(q: SelectQ): Promise<string>;
  select(q: string, options: string[], id: string, help?: string | null): Promise<string>;
  async select(q: SelectQ | string, options?: string[], id?: string, help?: string | null): Promise<string> {
    if (typeof q === 'string') {
      if (options == null) {
        throw new Error('options cannot be null');
      }
      if (id == null) {
        throw new Error('id cannot be null');
      }
      return this.select({
        type: QuestionType.Select,
        options,
        message: q,
        desc: help,
        id,
      });
    }
    const answer = this.getAnswer(q);

    if (isStringAnswer(answer)) return Promise.resolve(answer.answer);
    throw new Error('Incorrect answer type. Expected: StringAnswer. Got: ' + typeof answer);
  }

  text(q: TextQ): Promise<string>;
  text(q: string, id: string, help?: string | null): Promise<string>;
  async text(q: TextQ | string, id?: string, help?: string | null): Promise<string> {
    if (typeof q === 'string') {
      if (id == null) {
        throw new Error('id cannot be null');
      }
      return this.text({
        type: QuestionType.Text,
        message: q,
        desc: help,
        id,
      });
    }
    const answer = this.getAnswer(q);

    if (isStringAnswer(answer)) return Promise.resolve(answer.answer);
    throw new Error('Incorrect answer type. Expected: StringAnswer. Got: ' + typeof answer);
  }

  dateSelect(q: DateQ): Promise<string>;
  dateSelect(q: string, id: string, help?: string | null): Promise<string>;

  async dateSelect(q: DateQ | string, id?: string, help?: string | null): Promise<string> {
    if (typeof q === 'string') {
      if (id == null) {
        throw new Error('id cannot be null');
      }
      return this.dateSelect({
        type: QuestionType.DateSelect,
        message: q,
        desc: help,
        id,
      });
    }
    const answer = this.getAnswer(q);
    if (isStringAnswer(answer)) return Promise.resolve(answer.answer);
    throw new Error('Incorrect answer type. Expected: StringAnswer. Got: ' + typeof answer);
  }
}

export { StatelessInquirer };
